bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthConnector.java

```diff
@@ -0,0 +1,369 @@
+
+    private Fields initFields(String... parameters) {
+        Fields fields = new Fields();
+
+        for (int i = 0; i < parameters.length; i += 2) {
+            if (i + 1 < parameters.length && parameters[i] != null && parameters[i + 1] != null) {
+                logger.debug("Oauth request parameter {}, value {}", parameters[i], parameters[i + 1]);
+                fields.add(parameters[i], parameters[i + 1]);
+            }
+        }
+        return fields;
+    }
+
+    private OAuthToken doRequest(final String grantType, HttpClient httpClient, final Request request, Fields fields)
+            throws OAuthError, OAuthException {
+
+        try {
+            final FormContentProvider entity = new FormContentProvider(fields);
+            final ContentResponse response = AccessController
+                    .doPrivileged((PrivilegedExceptionAction<ContentResponse>) () -> {
+                        Request requestWithContent = request.content(entity);
+                        return requestWithContent.send();
+                    });
+
+            int statusCode = response.getStatus();
+            String content = response.getContentAsString();
+
+            if (statusCode == HttpStatus.OK_200) {
+                OAuthToken jsonResponse = gson.fromJson(content, OAuthToken.class);
+                logger.info("grant type {} to {} succeeded", grantType, request.getURI());
+                return jsonResponse;
+            } else if (statusCode == HttpStatus.BAD_REQUEST_400) {
+                OAuthError errorResponse = gson.fromJson(content, OAuthError.class);
+                logger.error("grant type {} to {} failed with error code {}, description {}", grantType,
+                        request.getURI(), errorResponse.getError(), errorResponse.getErrorDescription());
+
+                throw errorResponse;
+            } else {
+                logger.error("grant type {} to {} failed with HTTP response code {}", grantType, request.getURI(),
+                        statusCode);
+                throw new OAuthException("Bad http response, http code " + statusCode);
+            }
+        } catch (PrivilegedActionException pae) {
+            throw new OAuthException("Exception while oauth communication, grant type " + grantType,
+                    pae.getException());
+        }
+    }
+
+    private HttpClient getHttpClient(String tokenUrl) throws OAuthException {
+        try {
+            final URI tokenUri = new URI(tokenUrl);
+
+            SslContextFactory sslContextFactory = null;
+            // Get the trust managers from SslService
+            if (customTrustManagers == null) {
+                sslContextFactory = new SslContextFactory(); // use default, jdk's truststore etc
+            } else {
+                sslContextFactory = initSslContextFactory();
+            }
+
+            // Instantiate HttpClient with the SslContextFactory
+            HttpClient httpClient = new HttpClient(sslContextFactory);
```

Comment:
The recently introduced [HttpClientFactory](https://github.com/eclipse/smarthome/blob/master/bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/HttpClientFactory.java) should be used